server/access/streamBackend.go

```diff
@@ -0,0 +1,221 @@
+/*
+ * Flow Emulator
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package access
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/onflow/flow-emulator/emulator"
+	"github.com/onflow/flow-go/engine/access/state_stream"
+	"github.com/onflow/flow-go/engine/access/state_stream/backend"
+	"github.com/onflow/flow-go/engine/common/rpc"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/storage"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/utils/logging"
+	"github.com/rs/zerolog"
+	"time"
+)
+
+type StateStreamBackend struct {
+	blockchain *emulator.Blockchain
+
+	log            zerolog.Logger
+	sendTimeout    time.Duration
+	responseLimit  float64
+	sendBufferSize int
+
+	getExecutionData GetExecutionDataFunc
+	getStartHeight   GetStartHeightFunc
+}
+
+func NewStateStreamBackend(blockchain *emulator.Blockchain, log zerolog.Logger) *StateStreamBackend {
+	return &StateStreamBackend{
+		blockchain:       blockchain,
+		log:              log,
+		sendTimeout:      state_stream.DefaultSendTimeout,
+		responseLimit:    state_stream.DefaultResponseLimit,
+		sendBufferSize:   state_stream.DefaultSendBufferSize,
+		getExecutionData: getExecutionDataFunc(blockchain),
+		getStartHeight:   getStartHeightFunc(blockchain),
+	}
+}
+
+var _ state_stream.API = &StateStreamBackend{}
+
+func getStartHeightFunc(blockchain *emulator.Blockchain) GetStartHeightFunc {
+	return func(blockID flow.Identifier, height uint64) (uint64, error) {
+		block, err := blockchain.GetBlockByID(blockID)
+		if err == nil {
```

Comment:
Maybe worth checking type of an error: 
```go
errors.Is(err, storage.ErrNotFound)
```

Reply:
is the logic here to try with blockID first then try with block height?

Reply:
I would say the logic is fine, but if GetBlockByID is not found then return the height, but if the GetBlockByID returns an unknown error then I believe we should return that error.